File: spec/java/src/io/kaitai/struct/specwrite/TestEofExceptionBytes.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ByteBufferKaitaiStream;
import io.kaitai.struct.KaitaiStream;
import io.kaitai.struct.RandomAccessFileKaitaiStream;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.EofExceptionBytes;

import java.io.File;
import java.io.RandomAccessFile;

import org.testng.Assert.ThrowingRunnable;
import org.testng.annotations.Test;
import org.testng.SkipException;
import static org.testng.Assert.assertEquals;

public class TestEofExceptionBytes extends CommonSpec {
    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return EofExceptionBytes.class;
    }

    @Override
    protected String getSrcFilename() {
        return "term_strz.bin";
    }

    @Override
    @Test
    protected void testReadWriteRoundtrip() throws Exception {
        throw new SkipException("cannot use roundtrip because parsing is expected to fail");
    }

    @Test
    public void testEofExceptionBytesBB() throws Exception {
        EofExceptionBytes r = getEofExceptionBytes();

        try (KaitaiStream io = new ByteBufferKaitaiStream(12)) {
            assertThrowsEofError(new ThrowingRunnable() {
                @Override
                public void run() throws Throwable {
                    r._write(io);
                }
            });
        }
    }

    @Test
    public void testEofExceptionBytesRAF() throws Exception {
        EofExceptionBytes r = getEofExceptionBytes();

        File file = new File(SCRATCH_DIR + "specwrite_TestEofExceptionBytes.bin");
        RandomAccessFile raf = new RandomAccessFile(file, "rw");
        raf.setLength(12);

        Throwable thr;
        try (KaitaiStream io = new RandomAccessFileKaitaiStream(raf)) {
            thr = expectThrowsEofError(new ThrowingRunnable() {
                @Override
                public void run() throws Throwable {
                    r._write(io);
                }
            });
        } finally {
            file.delete();
        }
        assertEquals(thr.getMessage(), "requested to write 13 bytes, but only 12 bytes left in the stream");
    }

    protected EofExceptionBytes getEofExceptionBytes() {
        EofExceptionBytes r = new EofExceptionBytes();
        r.setBuf(new byte[] { 120, 121, 122, 123, 124, 125, 126, 127, -1, -2, -3, -4, -5 });
        r._check();

        return r;
    }
}

File: spec/python/specwrite/test_eof_exception_bytes.py
import unittest
import io
from kaitaistruct import KaitaiStream
from specwrite.common_spec import CommonSpec

from testwrite.eof_exception_bytes import EofExceptionBytes

class TestEofExceptionBytes(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super(TestEofExceptionBytes, self).__init__(*args, **kwargs)
        self.struct_class = EofExceptionBytes
        self.src_filename = 'src/term_strz.bin'

    def test_read_write_roundtrip(self):
        self.skipTest("cannot use roundtrip because parsing is expected to fail")

    def test_eof_exception_bytes(self):
        r = EofExceptionBytes()
        r.buf = b"\x78\x79\x7A\x7B\x7C\x7D\x7E\x7F\xFF\xFE\xFD\xFC\xFB"
        r._check()

        with KaitaiStream(io.BytesIO(bytearray(12))) as out_io:
            with self.assertRaisesRegex(EOFError, u"^requested to write 13 bytes, but only 12 bytes left in the stream$"):
                r._write(out_io)
